backend/LexBoxApi/GraphQL/ProjectMutations.cs

```diff
@@ -87,6 +96,64 @@ public record CreateProjectResponse(Guid? Id, CreateProjectResult Result);
         return dbContext.Projects.Where(p => p.Id == input.ProjectId);
     }
 
+    [Error<NotFoundException>]
+    [Error<DbError>]
+    [AdminRequired]
+    [UseMutationConvention]
+    public async Task<BulkAddProjectMembersResult> BulkAddProjectMembers(
+        LoggedInContext loggedInContext,
+        BulkAddProjectMembersInput input,
+        LexBoxDbContext dbContext)
+    {
+        var admin = await dbContext.Users.FindAsync(loggedInContext.User.Id);
+        var project = await dbContext.Projects.FindAsync(input.ProjectId);
+        if (project is null) throw new NotFoundException("Project not found");
+        List<string> usernameConflicts = [];
+        int createdCount = 0;
+        foreach (var username in input.Usernames)
+        {
+            var user = await dbContext.Users.Where(u => u.Username == username).FirstOrDefaultAsync();
+            if (user is null)
+            {
+                createdCount++;
+                var salt = Convert.ToHexString(RandomNumberGenerator.GetBytes(SHA1.HashSizeInBytes));
+                var isEmailAddress = username.Contains('@');
+                // TODO: In the future we'll want to allow usernames in the form "Real Name <email@example.com>" and extract the real name from them
+                // For now, just:
+                var name = username;
+                user = new User
+                {
+                    Id = Guid.NewGuid(),
+                    Username = isEmailAddress ? null : username,
+                    Name = name,
+                    Email = isEmailAddress ? username : null,
+                    LocalizationCode = "en", // TODO: input.Locale,
+                    Salt = salt,
+                    PasswordHash = PasswordHashing.HashPassword(input.PasswordHash, salt, true),
+                    IsAdmin = false,
+                    EmailVerified = false,
+                    CreatedBy = admin,
+                    Locked = false,
+                    CanCreateProjects = false
+                };
+                user.Projects.Add(new ProjectUsers { Role = input.Role, ProjectId = input.ProjectId, UserId = user.Id });
+                dbContext.Add(user);
+            }
+            else
+            {
+                usernameConflicts.Add(username);
+                var projectUser = await dbContext.ProjectUsers.FirstOrDefaultAsync(
+                    u => u.ProjectId == input.ProjectId && u.UserId == user.Id);
+                if (projectUser is null)
+                {
+                    user.Projects.Add(new ProjectUsers { Role = input.Role, ProjectId = input.ProjectId, UserId = user.Id });
+                }
```

Comment:
`dbContext.ProjectUsers` is a `Set`.
Can we just always do: `dbContext.ProjectUsers.Add(new ProjectUsers { Role = input.Role, ProjectId = input.ProjectId, UserId = user.Id })`?

Reply:
I seem to recall that I did that first and got duplicate key errors, but I'll try changing it to that and see what happens.

Reply:
No, we can't do that. It may be called a DbSet, but it's only acting as a set for the entities already loaded; it doesn't deduplicate things for you if it doesn't know about them. If we just do `dbContext.ProjectUsers.Add(new ProjectUsers { ... })` blindly, we'll get duplicate key errors. The [DbSet.Add docs](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbset-1.add?view=efcore-8.0#microsoft-entityframeworkcore-dbset-1-add(-0)) confirm that behavior.